### ISSUE TITLE
♻️TinaCMS - Remove lodash dependency

### ISSUE THREAD

**Issue by Ben0189** [tinacms/tinacms#4757]
[Lodash ](https://www.npmjs.com/package/lodash)packages is no longer updated (latest published 4 years ago) (it was introduced to provide some utility function for JavaScript project but Javascript has improved alot and we dont need lodash anymore, see this for some useful [resources](https://github.com/you-dont-need-x/you-dont-need-lodash)). These are just extra dependency to maintain.

Currently there are couple of packages in our monorepos that still using lodash, you might need to remove/replace all of them (graphql packages might be the tricky part because the code is hard to read)

- [x] Review the usage of lodash in each packages
- [ ] Replace lodash usage with something else (custom function if possible)
- [ ] Make sure the original function still working as it was before
- [ ] Remove lodash packages 

Something interesting to look at https://github.com/radashi-org/radashi

**Comment by nicko-winner** [comment]
https://es-toolkit.slash.page/intro.html offers a compatibility layer to lodash, which might be worth looking at: https://es-toolkit.slash.page/compatibility.html

**Comment by 18-th** [comment]
Direct `lodash` usage has been removed in #6145. Usages have been replaced with either native or `es-toolkit` alternatives.

`yup` and `jest-dom` dependency on lodash removed in #6187.

Transitive dependencies investigation:

# Lodash Dependencies Analysis

This document outlines all lodash dependencies in the TinaCMS project and the path to eliminating them.

## Dependency Graph

```mermaid
graph TD
    subgraph "Your Packages"
        tinacms["tinacms"]
        next-cloudinary["next-tinacms-cloudinary"]
        tinacms-clerk["tinacms-clerk"]
        tinacms-mdx["@tinacms/mdx"]
        schema-tools["@tinacms/schema-tools"]
    end

    subgraph "Lodash Variants"
        lodash["lodash"]
        lodash-es["lodash-es"]
        lodash-startcase["lodash.startcase"]
        lodash-memoize["lodash.memoize"]
        lodash-mapvalues["lodash.mapvalues"]
        lodash-sortby["lodash.sortby"]
        lodash-truncate["lodash.truncate"]
    end

    subgraph "Intermediate Dependencies"
        react-color["✅ react-color"]
        reactcss["reactcss"]
        cloudinary["🔇 cloudinary"]
        cloudinary-core["cloudinary-core"]
        mermaid["⚠️ mermaid"]
        dagre-d3-es["dagre-d3-es"]
        plate["⚠️ @udecode/plate"]
        slate-react["slate-react"]
        slate-dom["slate-dom"]
        zustand-x["zustand-x"]
        clerk-js["🔇 @clerk/clerk-js"]
        browser-tabs-lock["browser-tabs-lock"]
        concat-md["✅ concat-md"]
        ts-jest["✅ ts-jest"]
        langium["langium"]
        chevrotain["chevrotain"]
        request-promise["request-promise-core"]
        vitest["✅ vitest"]
        jsdom["jsdom"]
        whatwg-url["whatwg-url@7"]
        eslint["eslint@7/8"]
        table["table"]
    end

    tinacms --> react-color
    tinacms --> mermaid
    tinacms --> plate
    tinacms --> vitest

    react-color --> lodash
    react-color --> lodash-es
    react-color --> reactcss
    reactcss --> lodash

    mermaid --> dagre-d3-es
    mermaid --> langium
    dagre-d3-es --> lodash-es
    langium --> chevrotain
    chevrotain --> lodash-es

    plate --> slate-react
    plate --> slate-dom
    plate --> zustand-x
    slate-react --> lodash
    slate-dom --> lodash
    zustand-x --> lodash-mapvalues

    vitest -.-> jsdom
    jsdom --> whatwg-url
    whatwg-url --> lodash-sortby

    next-cloudinary --> cloudinary
    cloudinary --> cloudinary-core
    cloudinary --> request-promise
    cloudinary-core --> lodash
    request-promise --> lodash

    tinacms-clerk --> clerk-js
    clerk-js --> browser-tabs-lock
    browser-tabs-lock --> lodash

    tinacms-mdx --> concat-md
    concat-md --> lodash-startcase

    schema-tools --> ts-jest
    ts-jest --> lodash-memoize

    eslint --> table
    table --> lodash-truncate

    classDef lodashNode fill:#ff6b6b,stroke:#c92a2a,color:#fff
    classDef directDep fill:#4dabf7,stroke:#1971c2,color:#fff
    classDef intermediateDep fill:#ffd43b,stroke:#fab005,color:#333
    classDef blockedDep fill:#ff922b,stroke:#e8590c,color:#fff
    classDef canFixDep fill:#51cf66,stroke:#2f9e44,color:#fff
    classDef ignoredDep fill:#adb5bd,stroke:#868e96,color:#fff

    class lodash,lodash-es,lodash-startcase,lodash-memoize,lodash-mapvalues,lodash-sortby,lodash-truncate lodashNode
    class tinacms,next-cloudinary,tinacms-clerk,tinacms-mdx,schema-tools directDep
    class reactcss,dagre-d3-es,slate-react,slate-dom,zustand-x,browser-tabs-lock,langium,chevrotain,request-promise,jsdom,whatwg-url,eslint,table intermediateDep
    class mermaid,plate blockedDep
    class react-color,ts-jest,vitest,concat-md canFixDep
    class cloudinary,cloudinary-core,clerk-js ignoredDep
```

**Legend:**
- 🔵 Blue = Your direct workspace packages
- 🟢 Green (✅) = Can be replaced to remove lodash
- 🟡 Yellow = Intermediate dependencies (can be updated/replaced)
- 🟠 Orange (⚠️) = Blocked - cannot be replaced or upgrade won't help
- ⬜ Gray (🔇) = Ignored - optional client dependency
- 🔴 Red = Lodash variants (target to eliminate)
- Dashed line = Optional dependency

## Summary Table

| Package | Lodash Variant | Used By | Upgrade Removes Lodash? | Action |
|---------|---------------|---------|:----------------------:|--------|
| ✅ `react-color` | lodash, lodash-es | tinacms | ✅ Yes (replace) | Replace with `react-colorful` |
| `reactcss` | lodash | react-color → tinacms | ✅ Yes | Resolved when replacing react-color |
| 🔇 `cloudinary` | lodash (via cloudinary-core) | next-tinacms-cloudinary | 🔇 Ignored | Optional client template dependency |
| 🔇 `cloudinary-core` | lodash (peer dep) | cloudinary | 🔇 Ignored | Optional client template dependency |
| ⚠️ `mermaid` | lodash-es (via dagre-d3-es) | tinacms | ❌ No (latest still uses lodash-es) | **BLOCKED** - Core functionality |
| ⚠️ `@udecode/plate` | lodash (via slate-react, slate-dom, zustand-x) | tinacms | ❌ No (plate-core still uses lodash) | **BLOCKED** - Core functionality |
| `slate-react` | lodash | @udecode/plate → tinacms | ❌ No (latest still uses lodash) | Wait for upstream fix |
| `slate-dom` | lodash | @udecode/plate → tinacms | ❌ No | Wait for upstream fix |
| `zustand-x` | lodash.mapvalues | @udecode/plate → tinacms | ❌ No | Wait for upstream fix |
| 🔇 `@clerk/clerk-js` | lodash (via browser-tabs-lock) | tinacms-clerk | 🔇 Ignored | Optional client integration |
| `browser-tabs-lock` | lodash | @clerk/clerk-js → tinacms-clerk | 🔇 Ignored | Optional client integration |
| ✅ `concat-md` | lodash.startcase | @tinacms/mdx | ✅ Yes (replace) | Replace with custom script |
| ✅ `ts-jest` | lodash.memoize | @tinacms/schema-tools | ✅ Yes (switch to vitest) | Migrate to vitest |
| `table` | lodash.truncate | eslint@7/8 (indirect) | ⚠️ Partial (eslint v9 uses lodash.merge) | Upgrade eslint to v9+ |
| ✅ `whatwg-url@7` | lodash.sortby | vitest (optional jsdom) | ✅ Yes (upgrade vitest + jsdom) | Upgrade vitest to 2.x |
| `chevrotain` | lodash-es (via langium) | mermaid → tinacms | ❌ No | Wait for upstream fix |
| `request-promise-core` | lodash | cloudinary | ✅ Yes | Resolved when replacing cloudinary |

## Upgrade Analysis

### Packages Where Upgrading WILL Remove Lodash ✅

| Package | Current | Replace With | Notes |
|---------|---------|--------------|-------|
| ✅ `react-color` | 2.19.3 | `react-colorful` | Zero dependencies, different API |
| ✅ `concat-md` | 0.5.1 | Custom script | Dev dependency, ~20 lines to replace |
| ✅ `ts-jest` | 29.x | `vitest` | Migrate @tinacms/schema-tools to vitest |
| ✅ `vitest` (old) | 0.32.4 | `vitest@2.x` | Upgrade in @tinacms/graphql removes jsdom@15 → lodash.sortby |

### Packages Where Upgrading WON'T Remove Lodash ❌

| Package | Current | Latest | Why No Improvement |
|---------|---------|--------|-------------------|
| ⚠️ `mermaid` | 9.3.0 | 11.7.0 | **Core functionality** - Still uses `dagre-d3-es` → lodash-es |
| ⚠️ `@udecode/plate` | 48.x | 49.x | **Core functionality** - `plate-core` still depends on lodash |
| `eslint` | 8.x | 9.x | Removed `table` (lodash.truncate) but added `lodash.merge` |

## Priority Actions

### High Priority - Will Eliminate Lodash ✅

1. **react-color** → Replace with [`react-colorful`](https://github.com/omgovich/react-colorful) (zero deps)
2. **concat-md** → Replace with custom script (just need `startCase` function)
3. **ts-jest** → Migrate `@tinacms/schema-tools` to vitest
4. **vitest** → Upgrade `@tinacms/graphql` from 0.32.4 to 2.x (removes old jsdom → lodash.sortby)

### Low Priority - Won't Eliminate Lodash ❌

5. **eslint** → v9 still uses lodash.merge

### 🔇 Ignored - Optional Client Dependencies

These are optional templates/integrations - the client chooses to use them:
- 🔇 `cloudinary` ← Used in `next-tinacms-cloudinary` template
- 🔇 `@clerk/clerk-js` ← Used in `tinacms-clerk` integration

### ⚠️ Blocked - Core Functionality (Cannot Replace)

These are essential to TinaCMS and must wait for upstream fixes:
- ⚠️ `@udecode/plate` ← Rich text editor - waiting for `slate-react`, `slate-dom`, `zustand-x` to drop lodash
- ⚠️ `mermaid` ← Diagram rendering - waiting for `dagre-d3-es`, `chevrotain` to drop lodash

---

*Updated: Fri 28 Nov 2025*

**Comment by kulesy** [comment]
Update on scope after the v4 package-split work (epic #6831):

- **Core** (`@tinacms/tinacms`) — should be able to drop the `lodash` warning entirely once the 13 → 7 package consolidation lands. Direct lodash usage is already gone via #6145 and #6187.
- **Non-core / peer packages** — will still carry transitive `lodash` via upstream dependencies (`mermaid`, `@udecode/plate`). We accept that until those upstreams drop it.

So the "done" line for this issue is: core package free of the lodash warning. Transitive lodash in non-core packages is out of scope here — could spin a separate tracker if it ever becomes a security concern again.

Re-labeling with `For 4.1` so it gets picked up alongside the package split.
